File: 46.permutations.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode id=46 lang=java
 *
 * [46] Permutations
 */

// @lc code=start
class Solution {
    List<List<Integer>> res;
    int N;
    int[] nums;
    boolean[] visited;
    public List<List<Integer>> permute(int[] nums) {
        res = new ArrayList<>();
        N = nums.length;
        visited = new boolean[N];
        this.nums = nums;
        dfs(new ArrayList<>());
        return res;
        
    }
    void dfs(List<Integer> path){
        if(path.size() == N){
            res.add(new ArrayList<>(path));
            return;
        }
        //顺序是 count 的 --> 所以要有 visited
        for(int i= 0; i < N; i++){
            if(visited[i])
                continue;
            visited[i] = true;
            path.add(nums[i]);
            dfs(path);
            path.remove(path.size()-1);
            visited[i] = false;
        }
    }
}
// @lc code=end

File: js/leetcode/js/46.permutations.js
/*
 * @lc app=leetcode id=46 lang=javascript
 *
 * [46] Permutations
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
// 也可以用个数组 记录 used
var permute = function (nums) {
    function dfs(nums, path, res) {
        if (path.length === nums.length) {
            // console.log("path", path)
            Array.of(...path)
            res.push(Array.of(...path))
            return;
        }
        for (let i = 0; i < nums.length; i++) {
            if (nums[i] !== null) {
                const store = nums[i];
                path.push(nums[i]);
                nums[i] = null;
                dfs(nums, path, res)
                nums[i] = store;
                path.pop()
            }
        }

    }
    const res = [];
    dfs(nums, [], res)
    return res;

};
// @lc code=end

